lxd/device/device.go

```diff
@@ -0,0 +1,128 @@
+package device
+
+import (
+	"fmt"
+
+	"github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/state"
+)
+
+// devTypes defines supported top-level device type creation functions.
+var devTypes = map[string]func(config.Device) device{
+	"nic": loadNICByType,
+}
+
+// ErrUnsupportedDevType is the error that occurs when an unsupported device type is created.
+var ErrUnsupportedDevType = fmt.Errorf("Unsupported device type")
+
+// Device represents a device that can be added to an instance.
+type Device interface {
+	// CanHotPlug returns true if device can be managed whilst instance is running.
+	// It also returns a slice of config fields that can be live updated. If only fields in this
+	// list have changed then Update() is called rather than triggering a device remove & add.
+	CanHotPlug() (bool, []string)
+
+	// Add performs any host-side setup when a device is added to an instance.
+	// It is called irrespective of whether the instance is running or not.
+	Add() error
+
+	// Start peforms any host-side configuration required to start the device for the instance.
+	// This can be when a device is plugged into a running instance or the instance is starting.
+	// Returns run-time configuration needed for configuring the instance with the new device.
+	Start() (*RunConfig, error)
+
+	// Update performs host-side modifications for a device based on the difference between the
+	// current config and previous config supplied as an argument. This called if the only
+	// config fields that have changed are supplied in the list returned from CanHotPlug().
+	// The function also accepts a boolean indicating whether the instance is running or not.
+	Update(config.Device, bool) error
```

Comment:
`Update(oldConfig config.Device, running bool) error` would be easier to understand